utils/frame/rpc/system/src/lib.rs

```diff
@@ -17,224 +17,245 @@
+		})?;
+		Ok(Encode::encode(&result).into())
+	}
+}
+
+#[async_trait]
+impl<Client, Pool, Fetcher, Block, AccountId, Index>
+	SystemRpcBackend<<Block as traits::Block>::Hash, AccountId, Index>
+	for SystemRpcBackendLight<Client, Pool, Fetcher, Block>
+where
+	Client: Send + Sync + 'static,
+	Client: HeaderBackend<Block>,
+	Pool: TransactionPool + 'static,
+	Fetcher: light::Fetcher<Block> + 'static,
+	Block: traits::Block,
+	AccountId: Clone + Display + Codec + Send + 'static,
+	Index: Clone + Display + Codec + Send + traits::AtLeast32Bit + 'static,
+{
+	async fn nonce(&self, account: AccountId) -> Result<Index, JsonRpseeError> {
 		let best_hash = self.client.info().best_hash;
 		let best_id = BlockId::hash(best_hash);
-		let future_best_header = future_header(&*self.remote_blockchain, &*self.fetcher, best_id);
-		let fetcher = self.fetcher.clone();
+		let best_header = future_header(&*self.remote_blockchain, &*self.fetcher, best_id)
+			.await
+			.map_err(|blockchain_err| CallError::from_std_error(blockchain_err))?
+			.ok_or_else(|| ClientError::UnknownBlock(format!("{}", best_hash)))
+			.map_err(|client_err| CallError::from_std_error(client_err))?;
 		let call_data = account.encode();
-		let future_best_header = future_best_header.and_then(move |maybe_best_header| {
-			ready(
-				maybe_best_header
-					.ok_or_else(|| ClientError::UnknownBlock(format!("{}", best_hash))),
-			)
-		});
-
-		let future_nonce = future_best_header.and_then(move |best_header| {
-			fetcher.remote_call(RemoteCallRequest {
+		let nonce = self
+			.fetcher
+			.remote_call(RemoteCallRequest {
 				block: best_hash,
 				header: best_header,
 				method: "AccountNonceApi_account_nonce".into(),
 				call_data,
 				retry_count: None,
 			})
-		});
+			.await
+			.map_err(|blockchain_err| CallError::from_std_error(blockchain_err))?;
 
-		let future_nonce = future_nonce.and_then(|nonce| async move {
-			Index::decode(&mut &nonce[..])
-				.map_err(|e| ClientError::CallResultDecode("Cannot decode account nonce", e))
-		});
-		let future_nonce = future_nonce.map_err(|e| RpcError {
-			code: ErrorCode::ServerError(Error::RuntimeError.into()),
-			message: "Unable to query nonce.".into(),
-			data: Some(format!("{:?}", e).into()),
-		});
+		let nonce: Index = Decode::decode(&mut &nonce[..])
+			.map_err(|codec_err| CallError::from_std_error(codec_err))?;
 
-		let pool = self.pool.clone();
-		future_nonce.map_ok(move |nonce| adjust_nonce(&*pool, account, nonce)).boxed()
+		Ok(adjust_nonce(&*self.pool, account, nonce))
 	}
 
-	fn dry_run(
+	async fn dry_run(
 		&self,
 		_extrinsic: Bytes,
 		_at: Option<<Block as traits::Block>::Hash>,
-	) -> FutureResult<Bytes> {
-		async {
-			Err(RpcError {
-				code: ErrorCode::MethodNotFound,
-				message: "Unable to dry run extrinsic.".into(),
-				data: None,
-			})
+	) -> Result<Bytes, JsonRpseeError> {
+		Err(CallError::Custom {
+			code: jsonrpsee::types::v2::error::METHOD_NOT_FOUND_CODE,
```

Comment:
> code: -32601 | Message: Method not found

nitpick: this is technically not according to the spec AFAIU

I would prefer just `Error::to_call_error(anyhow::anyhow!("Not implemented for light clients").into())` the error code doesn't probably matter in this case....